tests/e2e/project.spec.ts

```diff
@@ -0,0 +1,337 @@
+    );
+  }
+});
+
+test("peer and branch switching", async ({ page }) => {
+  await page.goto(sourceBrowsingFixture);
+
+  // Alice's peer.
+  {
+    await page.getByTitle("Change peer").click();
+    await page.locator("text=alice").click();
+    await expect(page.getByTitle("Change peer")).toHaveText("alice delegate");
+    await expect(
+      page.locator("text=source-browsing / hyn1mj…qx7bun"),
+    ).toBeVisible();
+
+    // Default `main` branch.
+    {
+      await expect(page.getByTitle("Current branch")).toContainText(
+        "main 530aabd",
+      );
+      await expectCounts({ commits: 7, contributors: 1 }, page);
+    }
+
+    // Feature branch with a slash in the name.
+    {
+      await page.getByTitle("Change branch").click();
+      await page.locator("text=feature/branch").click();
+
+      await expect(page.getByTitle("Current branch")).toContainText(
+        "feature/branch d6318f7",
+      );
+      await expectCounts({ commits: 10, contributors: 1 }, page);
+    }
+
+    // Branch without a history or files in it.
+    {
+      await page.getByTitle("Change branch").click();
+      await page.locator("text=orphaned-branch").click();
+
+      await expect(page.getByTitle("Current branch")).toContainText(
+        "orphaned-branch af3641c",
+      );
+      await expectCounts({ commits: 1, contributors: 1 }, page);
+
+      await expect(
+        page.locator("text=We couldn't find any files at this revision."),
+      ).toBeVisible();
+    }
+  }
+
+  // Reset the source browser by clicking the project title.
+  {
+    await page.locator("text=source-browsing").click();
+
+    await expect(page.getByTitle("Change peer")).not.toContainText("alice");
+    await expect(page.getByTitle("Change peer")).not.toContainText("bob");
+
+    await expect(page.getByTitle("Current branch")).toContainText(
+      "main 530aabd",
+    );
+    await expect(page.locator("text=Git test repository")).toBeVisible();
+  }
+
+  // Bob's peer.
+  {
+    await page.getByTitle("Change peer").click();
+    await page.locator("text=bob").click();
+
+    await expect(page.getByTitle("Change peer")).toHaveText("bob");
+    await expect(page.getByTitle("Change peer")).not.toHaveText("delegate");
+
+    // Default `main` branch.
+    {
+      await expect(page.getByTitle("Current branch")).toContainText(
+        "main 0be0f03",
+      );
+      await expectCounts({ commits: 8, contributors: 2 }, page);
+      await expect(page.locator("text=0be0f03 Update readme")).toBeVisible();
+    }
+  }
+});
+
+test("clone modal", async ({ page }) => {
+  await page.goto(sourceBrowsingFixture);
+
+  await page.getByText("Clone").click();
+  await expect(
+    page.locator(
+      "text=rad clone rad://0.0.0.0/hnrkgd7sjt79k4j59ddh11ooxg18rk7soej8o",
+    ),
+  ).toBeVisible();
+  await expect(
+    page.locator(
+      "text=https://0.0.0.0/hnrkgd7sjt79k4j59ddh11ooxg18rk7soej8o.git",
+    ),
+  ).toBeVisible();
+});
```

Comment:
Maybe we can also test here the clipboard functionality?

Reply:
I also thought of testing clipboard functionality, but then decided against it, because there is no isolation for clipboard stuff and since tests run in parallel, there might be race conditions. I'll think on it. Maybe we can add all clipboard tests in one test to work around this issue.

Reply:
f9e5de89c26cc8fc2c60caf2a795f75d14971d64